File: solutions/907-koko-eating-bananas/koko-eating-bananas.rb
def min_eating_speed(piles, h)
    left = 1
    right = piles.max
    
    while left < right
        mid = left + (right - left) / 2
        
        if can_eat_all(piles, h, mid)
            right = mid
        else
            left = mid + 1
        end
    end
    
    left
end

def can_eat_all(piles, h, k)
    hours = 0
    piles.each do |pile|
        hours += (pile + k - 1) / k
        return false if hours > h
    end
    true
end

File: solutions/907-koko-eating-bananas/koko-eating-bananas.ts
function minEatingSpeed(piles: number[], h: number): number {
    let left: number = 1;
    let right: number = Math.max(...piles);
    
    while (left < right) {
        const mid: number = left + Math.floor((right - left) / 2);
        
        if (canEatAll(piles, h, mid)) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    
    return left;
}

function canEatAll(piles: number[], h: number, k: number): boolean {
    let hours: number = 0;
    for (const pile of piles) {
        hours += Math.ceil(pile / k);
        if (hours > h) return false;
    }
    return true;
}